### ISSUE TITLE
pedestrian dwell effect

### ISSUE THREAD

**Issue by jling888** [jling888/blockology-gvi#1]
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

def calculate_street_interface_matrix(
    vlm_pixel_df: pd.DataFrame,
    weights: dict = None,
    soft_barrier_permeability: float = 0.5,
) -> pd.DataFrame:
    """Computes the Street Interface Matrix (SIM) scores and sub-indices from VLM pixel classification data at 20m sampling nodes.

    Parameters:
    ------------
    vlm_pixel_df : pd.DataFrame
        DataFrame containing raw pixel counts per panorama.
        Required columns:
            - 'P_total': Total image resolution pixel count.
            - 'P_foliage_eye': Low-tier greenness (<2m height).
            - 'P_hard_barrier': Hard barriers (bollards, guardrails, walls).
            - 'P_soft_barrier': Soft barriers (hedges, linear planters).
            - 'P_shelter': Transitional shelters (arcades, overhangs, awnings).
            - 'P_building': Building facade pixels.
            - 'P_sky': Sky pixels.
            - 'P_rest': Micro-rest infrastructure (benches, ledges, stoops).
            - 'P_sidewalk': Accessible sidewalk surface pixels.
    weights : dict, optional
        Structural weights (alpha, beta, gamma, delta) summing to 1.0.
        Default values: alpha=0.30, beta=0.20, gamma=0.25, delta=0.25.
    soft_barrier_permeability : float
        Permeability coefficient (w_s) for soft vegetative buffers, in range (0, 1].

    Returns:
    --------
    results_df : pd.DataFrame
        DataFrame with computed raw sub-indices, normalized sub-indices, and final SIM scores.
    """
    df = vlm_pixel_df.copy()

    # 1. Default structural weights (summing to 1.0) if not supplied
    if weights is None:
        weights = {"alpha": 0.30, "beta": 0.20, "gamma": 0.25, "delta": 0.25}

    w_s = soft_barrier_permeability

    # 2. Compute Raw Sub-Indices

    # Eye-Level Green View Index (GVI_eye)
    df["GVI_eye_raw"] = df["P_foliage_eye"] / df["P_total"]

    # Edge Barrier Coefficient (EBC)
    df["EBC_raw"] = (
        df["P_hard_barrier"] + (w_s * df["P_soft_barrier"])
    ) / df["P_total"]

    # Transitional Enclosure Factor (TEF) - Normalized by Building + Sky openness
    df["TEF_raw"] = df["P_shelter"] / (df["P_building"] + df["P_sky"])

    # Spatial Affordance Index (SAI) - Normalized by accessible sidewalk area
    # Avoid division by zero with small epsilon
    epsilon = 1e-6
    df["SAI_raw"] = df["P_rest"] / (df["P_sidewalk"] + epsilon)

    # 3. Min-Max Normalization of Sub-Indices to [0, 1] Bounds
    scaler = MinMaxScaler()
    raw_cols = ["GVI_eye_raw", "EBC_raw", "TEF_raw", "SAI_raw"]
    norm_cols = ["GVI_eye_norm", "EBC_norm", "TEF_norm", "SAI_norm"]

    df[norm_cols] = scaler.fit_transform(df[raw_cols])

    # 4. Compute Composite Street Interface Matrix (SIM) Score
    df["SIM_score"] = (
        (weights["alpha"] * df["GVI_eye_norm"])
        + (weights["beta"] * df["EBC_norm"])
        + (weights["gamma"] * df["TEF_norm"])
        + (weights["delta"] * df["SAI_norm"])
    )

    return df

# =====================================================================
# EXAMPLE USAGE WITH SYNTHETIC VLM PIXEL DATASET
# =====================================================================
if __name__ == "__main__":
    np.random.seed(42)
    num_samples = 10

    # Simulate raw pixel outputs from a 2048x1024 resolution panorama (P_total = 2,097,152)
    P_tot = 2048 * 1024

    sample_vlm_data = pd.DataFrame(
        {
            "node_id": [f"SVI_{i:03d}" for i in range(num_samples)],
            "P_total": P_tot,
            "P_foliage_eye": np.random.randint(50000, 250000, size=num_samples),
            "P_hard_barrier": np.random.randint(10000, 80000, size=num_samples),
            "P_soft_barrier": np.random.randint(20000, 100000, size=num_samples),
            "P_shelter": np.random.randint(30000, 180000, size=num_samples),
            "P_building": np.random.randint(400000, 900000, size=num_samples),
            "P_sky": np.random.randint(200000, 600000, size=num_samples),
            "P_rest": np.random.randint(5000, 60000, size=num_samples),
            "P_sidewalk": np.random.randint(150000, 450000, size=num_samples),
        }
    )

    # Execute SIM Score Computation
    sim_results = calculate_street_interface_matrix(
        vlm_pixel_df=sample_vlm_data,
        weights={"alpha": 0.35, "beta": 0.15, "gamma": 0.25, "delta": 0.25},
        soft_barrier_permeability=0.6,
    )

    # Output Selected Results
    print("--- Street Interface Matrix (SIM) Computation Complete ---")
    print(
        sim_results[
            [
                "node_id",
                "GVI_eye_norm",
                "EBC_norm",
                "TEF_norm",
                "SAI_norm",
                "SIM_score",
            ]
        ]
    )
